### PR TITLE
Linkzorz

### PR DESCRIPTION
**Pull request by makosblade**
Make them link, make them think

First pass

### CODE REVIEW

**Comment by makosblade** [comment]
Pliz double-check my `SQL`z

File: src/main/kotlin/com/openlattice/linking/PostgresLinkingLogService.kt [hunk 1]
```diff
@@ -73,11 +71,29 @@ class PostgresLinkingLogService( val hds: HikariDataSource ) : LinkingLogService
...
+// Read latest version of a link
+private val READ_LATEST_LINKED_SQL = "SELECT ${ID_MAP.name} FROM ${LINKING_LOG.name} WHERE ${LINKING_ID.name} = ? ORDER BY ${VERSION.name} LIMIT 1"
+
+// Add a link. Grab previous value from a temp table
+private val ADD_LINK_SQL = "WITH old_json as " +
+        "( SELECT ${ID_MAP.name} as value FROM ${LINKING_LOG.name} WHERE ${LINKING_ID.name} = ? ORDER BY ${VERSION.name} LIMIT 1 )" +
+        "UPDATE ${LINKING_LOG.name} " +
+        "SET ${LINKING_ID.name} = ?" +
+        "${ID_MAP.name} = jsonb_set( old_json.value, '{?}', ${ID_MAP.name}->'?' || '?' )," +
+        "${VERSION.name} = ?" +
```
Comment (you-leee) [inline src/main/kotlin/com/openlattice/linking/PostgresLinkingLogService.kt]: Some spaces are missing from here

Comment (makosblade) [reply src/main/kotlin/com/openlattice/linking/PostgresLinkingLogService.kt]: Good catch! I was just writing some tests and saw that!

File: src/main/kotlin/com/openlattice/linking/PostgresLinkingLogService.kt [hunk 2]
```diff
@@ -73,11 +69,29 @@ class PostgresLinkingLogService( val hds: HikariDataSource ) : LinkingLogService
...
+        "UPDATE ${LINKING_LOG.name} " +
+        "SET ${LINKING_ID.name} = ?" +
+        "${ID_MAP.name} = jsonb_set( old_json.value, '{?}', ${ID_MAP.name}->'?' || '?'::jsonb )," +
+        "${VERSION.name} = ?" +
+        "FROM old_json" +
+        "WHERE ${LINKING_ID.name} = ?"
+
+// Delete Entity
+private val REMOVE_ENTITY_SQL= "UPDATE ${LINKING_LOG.name} " +
+        "SET ${ID_MAP.name} = ( $READ_LATEST_LINKED_SQL  #- '{?}')" +
```
Comment (you-leee) [inline src/main/kotlin/com/openlattice/linking/PostgresLinkingLogService.kt]: I'm not sure about this, but shouldn't removing entities also set version?

File: src/main/kotlin/com/openlattice/linking/PostgresLinkingLogService.kt [hunk 3]
```diff
@@ -73,11 +74,26 @@ class PostgresLinkingLogService( val hds: HikariDataSource ) : LinkingLogService
 
 private val LOG_COLUMNS = LINKING_LOG.columns.joinToString(",", transform = PostgresColumnDefinition::getName)
 
-private val INSERT_LOG_SQL = "INSERT INTO ${LINKING_LOG.name} ($LOG_COLUMNS) VALUES (?,?,?,?)"
+private val INSERT_LOG_SQL = "INSERT INTO ${LINKING_LOG.name} ($LOG_COLUMNS) VALUES (?,?::jsonb,?)"
+
+// Read latest version of a link
+private val READ_LATEST_LINKED_SQL = "SELECT ${ID_MAP.name} FROM ${LINKING_LOG.name} WHERE ${LINKING_ID.name} = ? ORDER BY ${VERSION.name} LIMIT 1"
```
Comment (katherinebernstein) [inline src/main/kotlin/com/openlattice/linking/PostgresLinkingLogService.kt]: does this need to be ordered desc?

File: src/main/kotlin/com/openlattice/linking/PostgresLinkingLogService.kt (hunk 2 above)
Comment (makosblade) [reply src/main/kotlin/com/openlattice/linking/PostgresLinkingLogService.kt]: yes! and it does w/new changes 🤦‍♂

File: src/main/kotlin/com/openlattice/linking/PostgresLinkingLogService.kt (hunk 3 above)
Comment (makosblade) [reply src/main/kotlin/com/openlattice/linking/PostgresLinkingLogService.kt]: yes! Brain works correctly today!

File: src/test/kotlin/com/openlattice/linking/PostgresLinkingLogServiceTest.kt [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+        private val objectMapper: ObjectMapper = ObjectMappers.getJsonMapper()
+        private val hds: HikariDataSource = HikariDataSource(HikariConfig(ConfigurationService.StaticLoader.loadConfiguration(RhizomeConfiguration::class.java)?.postgresConfiguration?.get()?.hikariConfiguration))
+        private var service: PostgresLinkingLogService = PostgresLinkingLogService(hds, objectMapper)
+
+
+        @BeforeClass
+        @JvmStatic
+        fun setUp() {
+            val sql = PostgresTable.LINKING_LOG.createTableQuery()
+            println("creating table $sql")
```
Comment (katherinebernstein) [inline src/test/kotlin/com/openlattice/linking/PostgresLinkingLogServiceTest.kt]: por qué no logger.info?

Comment (makosblade) [reply src/test/kotlin/com/openlattice/linking/PostgresLinkingLogServiceTest.kt]: 🤦‍♂ 😪

**Pull request merged**